commons/ava_testsuite/staking_gecko_network_tests.go

```diff
@@ -191,21 +179,266 @@ func (test FiveNodeStakingNetworkGetValidatorsTest) GetTimeout() time.Duration {
+			acceptableNodeIds[badServiceNodeId2] = true
+			verifyExpectedPeers(context, serviceId, allGeckoClients[serviceId], acceptableNodeIds, len(allServiceIds) - 1, false)
+		} else {
+			verifyExpectedPeers(context, serviceId, allGeckoClients[serviceId], acceptableNodeIds, len(bootServiceIds), false)
+		}
+	}
+	logrus.Info("Verified that the network has settled on the second node with previously-duplicated ID")
+}
+
+func (f FiveNodeStakingNetworkDuplicateIdTest) GetNetworkLoader() (testsuite.TestNetworkLoader, error) {
+	return getFiveNodeStakingLoader()
+}
+
+func (f FiveNodeStakingNetworkDuplicateIdTest) GetTimeout() time.Duration {
+	return 120 * time.Second
+}
+
 // =============== Helper functions =============================
 
 // TODO TODO TODO Rename this
 func getFiveNodeStakingLoader() (testsuite.TestNetworkLoader, error) {
 	serviceConfigs := map[int]ava_networks.TestGeckoNetworkServiceConfig{
-		NODE_CONFIG_ID: *ava_networks.NewTestGeckoNetworkServiceConfig(true, ava_services.LOG_LEVEL_DEBUG),
+		NORMAL_NODE_CONFIG_ID: *ava_networks.NewTestGeckoNetworkServiceConfig(true, ava_services.LOG_LEVEL_DEBUG),
+		SAME_CERT_CONFIG_ID:   *ava_networks.NewTestGeckoNetworkServiceConfig(false, ava_services.LOG_LEVEL_DEBUG),
 	}
 	return ava_networks.NewTestGeckoNetworkLoader(
 		ava_services.LOG_LEVEL_DEBUG,
 		true,
 		serviceConfigs,
 		map[int]int{
-			NODE_SERVICE_ID: NODE_CONFIG_ID,
+			NODE_SERVICE_ID: NORMAL_NODE_CONFIG_ID,
 		},
 		2,
 		2)
 }
 
+/*
+Verifies that a node's actual peers are what we expect
+
+Args:
+	context: Test context (used for failing if there's a problem)
+	serviceId: Service ID of the node whose peers are being examined
+	client: Gecko client for the node being examined
+	acceptableNodeIds: A "set" of acceptable node IDs where, if a peer doesn't have this ID, the test will be failed
+	expectedNumPeers: The number of peers we expect this node to have
+	atLeast: If true, indicates that the number of peers must be AT LEAST the expected number of peers; if false, must be exact
+ */
+func verifyExpectedPeers(
+			context testsuite.TestContext,
+			serviceId int,
+			client *gecko_client.GeckoClient,
+			acceptableNodeIds map[string]bool,
+			expectedNumPeers int,
+			atLeast bool) {
+	peers, err := client.AdminApi().GetPeers()
+	if err != nil {
+		context.Fatal(stacktrace.Propagate(err, "Failed to get peers from service with ID %v", serviceId))
+	}
+
+	actualNumPeers := len(peers)
+	var condition bool
+	var operatorAsserted string
+	if atLeast {
+		condition = actualNumPeers >= expectedNumPeers
+		operatorAsserted = ">="
+	} else {
+		condition = actualNumPeers == expectedNumPeers
+		operatorAsserted = "=="
+	}
+	context.AssertTrue(
+		condition,
+		stacktrace.NewError(
+			"Service ID %v actual num peers, %v, is not %v expected num peers, %v",
+			serviceId,
+			actualNumPeers,
+			operatorAsserted,
+			expectedNumPeers,
+		),
+	)
+
+	// Verify that IDs of the peers we have are in our list of acceptable IDs
```

Comment:
in fully connected test on master, we actually iterate through the acceptableNodeIds and make sure they're all in the peer list, _except_ the nodeID of the node itself (the node represented by the client arg to this function)

Reply:
yeah so the reason I can't do this here is for the same reason as above - in the dupe node ID test, you don't actually know what the peer list will look like

Reply:
so if i understand this loop correctly, every peer in the client's peer list must show up in the acceptableNodeIds map for it to pass

Reply:
yep correct - peers must be a subset of acceptableNodeIds

Reply:
i seee i see. cool. heads up, for bootstrappers, acceptableNodeIDs is actually a subset of peers. wild. not sure if there's a way to configure both verifications in one function with a flag